File: CaddieAI/Models/HoleAnalysis.swift
//
//  HoleAnalysis.swift
//  CaddieAI
//
//  Result model for hole geometry analysis and strategic advice.
//

import Foundation

// MARK: - Hole Analysis Result

struct HoleAnalysis: Codable, Sendable {
    var holeNumber: Int
    var par: Int?
    var totalDistanceYards: Int?
    var yardagesByTee: [String: Int]?
    var dogleg: DoglegInfo?
    var fairwayWidthAtLandingYards: Int?
    var greenDepthYards: Int?
    var greenWidthYards: Int?
    var hazards: [HoleHazardInfo]
    var strategicAdvice: String?
    var deterministicSummary: String
}

// MARK: - Dogleg Info

struct DoglegInfo: Codable, Sendable {
    var direction: DoglegDirection
    var distanceFromTeeYards: Int
    var bendAngleDegrees: Double
}

enum DoglegDirection: String, Codable, Sendable {
    case left, right

    var displayName: String {
        switch self {
        case .left: return "left"
        case .right: return "right"
        }
    }
}

// MARK: - Hazard Info

struct HoleHazardInfo: Codable, Sendable {
    var type: HazardType
    var side: HazardSide
    var distanceFromTeeYards: Int?
    var description: String
}

enum HazardType: String, Codable, Sendable {
    case water, bunker

    var displayName: String {
        switch self {
        case .water: return "Water"
        case .bunker: return "Bunker"
        }
    }
}

enum HazardSide: String, Codable, Sendable {
    case left, right, crossing, greenside, frontOfGreen

    var displayName: String {
        switch self {
        case .left: return "Left"
        case .right: return "Right"
        case .crossing: return "Crossing"
        case .greenside: return "Greenside"
        case .frontOfGreen: return "Front of green"
        }
    }
}

File: lib/core/golf/hole_analysis.dart
// HoleAnalysis output types — KAN-293 (S7.2). Direct port of the
// supporting structs in `ios/CaddieAI/Services/HoleAnalysisEngine.swift`.
//
// `HoleAnalysis` is the deterministic geometric analysis of a
// single hole: total length, dogleg info (if any), fairway width,
// green dimensions, classified hazards, and a short text summary
// the LLM router (KAN-294) consumes as prompt context.

enum DoglegDirection { left, right }

enum HazardType { bunker, water }

enum HazardSide { left, right, frontOfGreen, greenside, crossing }

class DoglegInfo {
  const DoglegInfo({
    required this.direction,
    required this.distanceFromTeeYards,
    required this.bendAngleDegrees,
  });

  final DoglegDirection direction;
  final int distanceFromTeeYards;
  final double bendAngleDegrees;
}

class GreenDimensions {
  const GreenDimensions({required this.depth, required this.width});

  /// Yards front-to-back along the approach axis.
  final int depth;

  /// Yards perpendicular to the approach axis.
  final int width;
}

class HoleHazardInfo {
  const HoleHazardInfo({
    required this.type,
    required this.side,
    required this.distanceFromTeeYards,
    required this.description,
  });

  final HazardType type;
  final HazardSide side;
  final int? distanceFromTeeYards;
  final String description;
}

class HoleAnalysis {
  const HoleAnalysis({
    required this.holeNumber,
    required this.par,
    required this.totalDistanceYards,
    required this.dogleg,
    required this.fairwayWidthAtLandingYards,
    required this.greenDepthYards,
    required this.greenWidthYards,
    required this.hazards,
    required this.deterministicSummary,
  });

  final int holeNumber;
  final int? par;
  final int? totalDistanceYards;
  final DoglegInfo? dogleg;
  final int? fairwayWidthAtLandingYards;
  final int? greenDepthYards;
  final int? greenWidthYards;
  final List<HoleHazardInfo> hazards;

  /// Short, plain-text summary the LLM router uses as additional
  /// context. Built by `HoleAnalysisEngine._buildSummary`.
  final String deterministicSummary;
}
